code/scorpio/troll.dm

```diff
@@ -0,0 +1,120 @@
+// troll.dm
+// Copyright 2021 Patrick Meade.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//------------------------------------------------------------------------------
+
+GLOBAL_LIST_EMPTY(gamer_words)
+
+/proc/add_gamer_words(config_value)
+	GLOB.gamer_words += splittext(config_value, ",")
+
+/proc/ban_and_kick_troll(context, message, word, mob_name, client/ban_me)
+	ban_troll(context, message, word, mob_name, ban_me)
+	kick_troll(ban_me)
+
+/proc/ban_troll(context, message, word, mob_name, client/ban_me)
```

Comment:
Why have this entire proc when you can just use the existing `DB_ban_record()`

Reply:
My first draft did exactly that, called `DB_ban_record()`, and I ran into a few issues.

* I needed a synthetic `/datum/admins` to represent the banning admin to call the `DB_ban_record` proc upon.
* The synthetic admin needed a `/client` because `DB_ban_record` is coded to read the admin's values (ckey, ip, cid) from the `owner` field (with its own special getter/setter procs that take a `/client`). Without a `/client`, empty values for fields are sent to NOT NULL columns in the SQL query.
    * Problem: BYOND will not let you instantiate a `/client` programmatically.
        * Workaround: Modify `DB_ban_record` to take admeme_ckey, admeme_ip, admeme_cid parameters.
* `DB_ban_record` uses `check_rights` to ensure an admin with authority to ban is the caller. This call doesn't use the rights assigned to the synthetic `/datum/admins`, but rather the rights assigned to `usr` (by default)
    * Problem: I want the ban to be triggered by a regular player's action; a self-ban that doesn't involve an admin
        * Potential Workaround: Modify `DB_ban_record` to take a parameter to bypass the rights check or check the passed in synthetic admin instead of the implicit actor.

At this point, I cut bait. Re-using the existing `DB_ban_record` would have been nice, but it was clearly designed to handle a very different use-case. Had I been renovating the entire ban system, I would have invested the effort.

So I gently borrowed the important parts of `DB_ban_record`, rolled it up in a new proc, and it worked like a charm.
Of course, now I've got another SQL query that needs to be updated if table `ban` ever changes.
But, I'm willing to take that compromise.

Reply:
> Potential Workaround: Modify DB_ban_record to take a parameter to bypass the rights check or check the passed in synthetic admin instead of the implicit actor.

`/datum/admins/proc/DB_ban_record(var/bantype, var/mob/banned_mob, var/duration = -1, var/reason, var/job = "", var/rounds = 0, var/banckey = null, var/banip = null, var/bancid = null)
`

to:

`/datum/admins/proc/DB_ban_record(var/bantype, var/mob/banned_mob, var/duration = -1, var/reason, var/job = "", var/rounds = 0, var/banckey = null, var/banip = null, var/bancid = null, var/rbypass = FALSE)`

?

Since optional variables are not required for every proc call, you would just need to put the code into `DB_ban_record` and, when, calling DB_ban_record from the Troll code, yes, fill in every variable as applicable with the last one being TRUE.

It's, uh, how I implemented my **\*cough\*** maybe not _great_ **\*cough\*** ANESTHESIA subsystem; by adding optional variables to Sleeping() and other procs.

Reply:
> I needed a synthetic /datum/admins to represent the banning admin to call the DB_ban_record proc upon.

Can it be put on the /datum level instead?

> The synthetic admin needed a /client

If so, this wouldn't be necessary

> if(!check_rights(R_BAN))	return

`if(!check_rights(R_BAN) && !rbypass)) return`

Reply:
Ha! Never mind! As I was typing this up, you were explaining it to me on Discord. @blinkdog

> Yeah, I wrestled with DB_ban_record for quite a bit before deciding not to use it
The main problem is a baked-in dependency on owner, a /client object, that only the BYOND engine itself is allowed to create.
The second problem is the way check_rights is called; it needs an optional argument to check the admin rights of another person
By default it checks the rights of the person taking the action; i.e.: the troll who is saying the n-word, who probably doesn't have admin rights to ban people
It could be fixed by adding four optional arguments, but it's already got 9 arguments;
making it 13 feels like an exercise in pain for someone trying to understand how that proc works